src/Microsoft.AspNetCore.Mvc.Core/Internal/ActionMethodExecutor.cs

```diff
@@ -0,0 +1,189 @@
+                }
+            }
+
+            Debug.Fail("Should not get here");
+            throw new Exception();
+        }
+
+        // void LogMessgae(..)
+        private class VoidResultExecutor : ActionMethodExecutor
+        {
+            public override ValueTask<IActionResult> Execute(ObjectMethodExecutor executor, object controller, object[] arguments)
+            {
+                executor.Execute(controller, arguments);
+                return new ValueTask<IActionResult>(new EmptyResult());
+            }
+
+            protected override bool CanConvert(ObjectMethodExecutor executor) 
+                => !executor.IsMethodAsync && executor.MethodReturnType == typeof(void);
+        }
+
+        // IActionResult Post(..)
+        // CreatedAtResult Put(..)
+        private class SyncActionResultExecutor : ActionMethodExecutor
+        {
+            public override ValueTask<IActionResult> Execute(ObjectMethodExecutor executor, object controller, object[] arguments)
+            {
+                var actionResult = (IActionResult)executor.Execute(controller, arguments);
+                EnsureActionResultNotNull(executor, actionResult);
+
+                return new ValueTask<IActionResult>(actionResult);
+            }
+
+            protected override bool CanConvert(ObjectMethodExecutor executor)
+                => !executor.IsMethodAsync && typeof(IActionResult).IsAssignableFrom(executor.MethodReturnType);
+        }
+
+        // Person GetPerson(..)
+        // object Index(..)
+        private class SyncObjectResultExecutor : ActionMethodExecutor
+        {
+            public override ValueTask<IActionResult> Execute(ObjectMethodExecutor executor, object controller, object[] arguments)
+            {
+                // Sync method returning arbitrary object
+                var returnValue = executor.Execute(controller, arguments);
+                var actionResult = returnValue as IActionResult ?? new ObjectResult(returnValue)
+                {
+                    DeclaredType = executor.MethodReturnType,
+                };
+
+                return new ValueTask<IActionResult>(actionResult);
+            }
+
+            // Catch-all for sync methods
+            protected override bool CanConvert(ObjectMethodExecutor executor) => !executor.IsMethodAsync;
+        }
+
+        // Task SaveState(..)
+        private class TaskResultExecutor : ActionMethodExecutor
+        {
+            public override async ValueTask<IActionResult> Execute(ObjectMethodExecutor executor, object controller, object[] arguments)
+            {
+                await (Task)executor.Execute(controller, arguments);
+                return new EmptyResult();
+            }
+
+            protected override bool CanConvert(ObjectMethodExecutor executor) => executor.MethodReturnType == typeof(Task);
+        }
+
+        // CustomAsync PerformActionAsync(..)
+        private class AsyncVoidResultExecutor : ActionMethodExecutor
```

Comment:
Can we call this 'awaitable'? I don't want to confuse this with `async void`. Also if we have this case do we need `TaskResultExecutor`?

Reply:
We can call `Execute` in `TaskResultExecutor` and avoid the cost of the `ExecuteAsync` pipeline.